### PR TITLE
Implement action key path to implicit declaration

### PR DESCRIPTION
**Pull request by Den-Rimus**
Support multiple key params
Rename annotation
Adjust sample events and tests

### CODE REVIEW

File: service-compiler/src/main/java/io/techery/analytics/compiler/ActionClassUtils.java [hunk 1]
```diff
@@ -89,6 +91,32 @@ public static String getAnnotatedFieldName(Elements elementUtils, TypeElement ty
       return attributes;
    }
 
+   public static Set<KeyPathEntity> getKeyPathEntities(Elements elementUtils, TypeElement typeElement) {
+      final Set<KeyPathEntity> keyPathEntities = new HashSet<KeyPathEntity>();
+      final String keyPathSymbol = "$"; // TODO make '$' customizable
```
Comment (almozavr) [inline service-compiler/src/main/java/io/techery/analytics/compiler/ActionClassUtils.java]: should be `const` to be easily identifiable durink quick class look through

File: service-compiler/src/main/java/io/techery/analytics/compiler/ActionClassUtils.java [hunk 2]
```diff
@@ -89,6 +91,32 @@ public static String getAnnotatedFieldName(Elements elementUtils, TypeElement ty
...
+               ));
+            }
+         }
+      }
+
+      return keyPathEntities;
+   }
+
+   public static boolean checkHasKeyParams(String actionKey) {
+      final String keyParamSymbol = "$"; // TODO: make this customizable via gradle
```
Comment (almozavr) [inline service-compiler/src/main/java/io/techery/analytics/compiler/ActionClassUtils.java]: same const

File: service-compiler/src/main/java/io/techery/analytics/compiler/ActionHelperGenerator.java [hunk 3]
```diff
@@ -79,8 +80,15 @@ private CodeBlock makeGetActionCodeBlock(AnalyticActionClass actionClass) {
       final CodeBlock.Builder builder = CodeBlock.builder();
 
       if (actionClass.containsActionPathParam) {
-         builder.addStatement("return $S.replace($S, action.$L)",
-               actionClass.action, ACTION_PATH_PARAM, actionClass.actionPartFieldName);
+         builder.addStatement("$T $L = $S", String.class, GENERATED_ACTION_KEY_VAR_NAME, actionClass.action);
+         for (KeyPathEntity keyPathEntity : actionClass.keyPathEntities) {
```
Comment (almozavr) [inline service-compiler/src/main/java/io/techery/analytics/compiler/ActionHelperGenerator.java]: in terms of performance, it's no so good to have multiple lines of string instances with replace method -> so better to replace actual string with it's parts and `+` signs

File: service-compiler/src/main/java/io/techery/analytics/compiler/model/AnalyticActionClass.java [hunk 4]
```diff
@@ -51,11 +50,10 @@ private void obtainAttributes(Elements elementUtils, TypeElement typeElement) {
       if (superTypeElement != null) {
          obtainAttributes(elementUtils, superTypeElement);
       }
+
       if (containsActionPathParam) {
-         if (actionPartFieldName == null) {
-            if (ActionClassUtils.checkHasAnnotatedField(elementUtils, typeElement, ActionPart.class)) {
-               actionPartFieldName = ActionClassUtils.getAnnotatedFieldName(elementUtils, typeElement, ActionPart.class);
-            }
+         if (ActionClassUtils.checkHasAnnotatedField(elementUtils, typeElement, KeyPath.class)) {
```
Comment (almozavr) [inline service-compiler/src/main/java/io/techery/analytics/compiler/model/AnalyticActionClass.java:55]: there should be validation or TODO for validation: if there are more pathes then `KeyPath` fields

**Review by almozavr — CHANGES_REQUESTED** [review]
Also I'd like to discuss replaces logic: do we allow multiple same-name key pathes? If no (I suppose) we should validate such case.

File: service-compiler/src/main/java/io/techery/analytics/compiler/ActionHelperGenerator.java (hunk 3 above)
Comment (Den-Rimus) [reply service-compiler/src/main/java/io/techery/analytics/compiler/ActionHelperGenerator.java]: Looking at effort needed to implement this I'd say that it does not worth it.
Please approve that this is really a blocker

Comment (almozavr) [reply service-compiler/src/main/java/io/techery/analytics/compiler/ActionHelperGenerator.java]: it's not a blocker, an improvement

**Pull request merged**